### PR TITLE
System labels for messages and calls

### PR DESCRIPTION
**Pull request by rowanseymour**
Not quite merge-able yet due to the Sent view being very slow for large orgs.

Replaces all of the OrgFolder counts in redis stuff

### CODE REVIEW

File: temba/api/serializers.py [hunk 1]
```diff
@@ -228,7 +228,8 @@ def restore_object(self, attrs, instance=None):
         msg_ids = attrs['messages']
         action = attrs['action']
 
-        msgs = Msg.objects.filter(org=self.org, direction=INCOMING, pk__in=msg_ids).select_related('contact')
+        msgs = Msg.objects.filter(org=self.org, direction=INCOMING, pk__in=msg_ids).exclude(visibility=DELETED)
```
Comment (rowanseymour) [inline temba/api/serializers.py]: Don't let API users act on deleted msgs

File: temba/api/tests.py [hunk 2]
```diff
@@ -2837,8 +2837,16 @@ def test_received(self):
         # should get 404 as the channel wasn't found
         self.assertEquals(404, response.status_code)
 
-        # change our message to incoming
-        sms.direction = 'I'
+    def test_delivered(self):
```
Comment (rowanseymour) [inline temba/api/tests.py]: @nicp I think we've made this same change in two branches - tests can no longer switch a msg direction

File: temba/msgs/migrations/0026_populate_system_labels.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        'X': Msg.objects.filter(direction='O', visibility='V', status='F').exclude(contact__is_test=True),
+        'E': Broadcast.objects.all().exclude(schedule=None).exclude(contacts__is_test=True),
+        'C': Call.objects.filter(is_active=True).exclude(contact__is_test=True)
+    }
+
+    for label_type, queryset in SYSLABEL_QUERYSETS.iteritems():
+        # items created after this time will have been already included via triggers
+        queryset = queryset.filter(created_on__lt=start_time)
+
+        # grab aggregate counts for all orgs - faster than doing org by org
```
Comment (rowanseymour) [inline temba/msgs/migrations/0026_populate_system_labels.py]: Takes ~6 mins on my machine doing it this way

File: temba/api/tests.py (hunk 2 above)
Comment (NicP) [inline temba/api/tests.py]: I think you misspelled the account name of whoever make the change you mentioned. I did not even remember I had this nicp account and I have not worked in Python for years... 

This means whoever the message was meant to reach did not receive it.

Comment (ericnewcomer) [inline temba/api/tests.py]: He sure did, he meant @nicpottier. Sorry about that! Thanks for the heads up.

File: temba/msgs/migrations/0027_create_sent_index.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import migrations
+
+# language=SQL
+CREATE_SQL = """
+CREATE INDEX msgs_msg_sent_label ON msgs_msg(org_id, created_on DESC)
+WHERE direction = 'O' AND visibility = 'V' AND status IN ('W', 'S', 'D');
```
Comment (rowanseymour) [inline temba/msgs/migrations/0027_create_sent_index.py]: This index is about ~1GB on rapidpro dump from June 5th

File: temba/contacts/models.py [hunk 5]
```diff
@@ -446,12 +446,6 @@ def get_or_create(cls, org, user, name=None, urns=None, incoming_channel=None, u
                 updated_attrs = dict(name=name, org=org, created_by=user, modified_by=user, is_test=is_test)
                 contact = Contact.objects.create(**updated_attrs)
 
-                # add it to our All Contacts group
-                if not contact.is_test:
```
Comment (rowanseymour) [inline temba/contacts/models.py]: This is handled by a trigger and should have been removed

File: temba/contacts/models.py [hunk 6]
```diff
@@ -1468,6 +1466,24 @@ def update_groups_for_contact(cls, contact, field=None):
...
 
+    @classmethod
+    def get_system_group_queryset(cls, org, group_type):
+        if group_type == cls.TYPE_USER_DEFINED:
+            raise ValueError("Can only get system group querysets")
+
+        return cls.all_groups.get(org=org, group_type=group_type).contacts.all()
+
+    @classmethod
+    def get_system_group_counts(cls, org, group_types=None):
```
Comment (rowanseymour) [inline temba/contacts/models.py]: For consistency with system labels - so you can grab all counts in one fetch

**Comment by rowanseymour** [comment]
@nicpottier @norkans7 please review

File: temba/contacts/migrations/0001_initial.py [hunk 7]
```diff
@@ -44,7 +44,7 @@ class Migration(migrations.Migration):
             ],
             options={
             },
-            bases=(models.Model, temba.orgs.models.OrgModelMixin),
```
Comment (nicpottier) [inline temba/contacts/migrations/0001_initial.py]: What's removing these about?

File: temba/msgs/models.py [hunk 8]
```diff
@@ -1373,25 +1367,115 @@ def create_call(cls, channel, phone, date, duration, call_type, user=None):
...
+        trigger used to maintain the label counts.
+        """
+        if label_type == cls.TYPE_INBOX:
+            qs = Msg.objects.filter(direction=INCOMING, visibility=VISIBLE, msg_type=INBOX)
+        elif label_type == cls.TYPE_FLOWS:
+            qs = Msg.objects.filter(direction=INCOMING, visibility=VISIBLE, msg_type=FLOW)
+        elif label_type == cls.TYPE_ARCHIVED:
+            qs = Msg.objects.filter(direction=INCOMING, visibility=ARCHIVED)
+        elif label_type == cls.TYPE_OUTBOX:
+            qs = Msg.objects.filter(direction=OUTGOING, visibility=VISIBLE, status__in=(PENDING, QUEUED))
```
Comment (nicpottier) [inline temba/msgs/models.py]: What's the performance of this query like for 600k queued messages?

File: temba/msgs/models.py [hunk 9]
```diff
@@ -1411,15 +1495,15 @@ class Label(TembaModel, SmartModel):
...
+    label_type = models.CharField(max_length=1, choices=TYPE_CHOICES, default=TYPE_LABEL, help_text=_("Label type"))
 
     visible_count = models.PositiveIntegerField(default=0,
                                                 help_text=_("Number of non-archived messages with this label"))
 
     # define some custom managers to do the filtering of label types for us
-    user_all = models.Manager()
-    user_folders = UserFolderManager()
-    user_labels = UserLabelManager()
+    objects = models.Manager()
```
Comment (nicpottier) [inline temba/msgs/models.py]: My bias for cases like this is to remove ```objects``` and use something more explicit, as it seems likely callers will just think of Labels as only being labels and not Folders. So having this be something like ```label_and_folder_objects``` even might drive that point home.

File: temba/contacts/migrations/0001_initial.py (hunk 7 above)
Comment (rowanseymour) [inline temba/contacts/migrations/0001_initial.py]: OrgModelMixin doesn't exist any longer

Comment (nicpottier) [inline temba/contacts/migrations/0001_initial.py]: Oh, a fine reason then!

File: temba/msgs/models.py (hunk 9 above)
Comment (rowanseymour) [inline temba/msgs/models.py]: `label_and_folder_objects`... has your account been hijacked by a J2EE developer :) Am not convinced the verbosity is worth it so tell me how strongly you feel about this before I change it.

Comment (nicpottier) [inline temba/msgs/models.py]: Hah. I do feel strongly it shouldn't be ```objects``` but you can call it ```funions``` as far as I'm concerned. The key thing is just that the default behavior of doing ```Label.objects``` doesn't work hopefully causing future RapidPro devs to wonder why and then pick the appropriate manager.

File: temba/contacts/models.py [hunk 10]
```diff
@@ -1313,31 +1311,31 @@ class Meta:
...
+        return super(UserContactGroupManager, self).get_queryset().filter(group_type=ContactGroup.TYPE_USER_DEFINED)
 
 
 class ContactGroup(TembaModel, SmartModel):
+    TYPE_ALL = 'A'
+    TYPE_BLOCKED = 'B'
+    TYPE_FAILED = 'F'
+    TYPE_USER_DEFINED = 'U'
+
+    TYPE_CHOICES = ((TYPE_ALL, "All Contacts"),
```
Comment (norkans7) [inline temba/contacts/models.py]: localization on the text

Comment (rowanseymour) [inline temba/contacts/models.py]: These are never exposed in the UI so I think it's better to keep these strings out of translation files. Actually ideally we might in the future follow the `Label` vs `SystemLabel` approach and move them out of `ContactGroup` altogether so their names don't conflict with user groups etc.

File: temba/msgs/models.py (hunk 9 above)
Comment (rowanseymour) [inline temba/msgs/models.py]: The managers are now: `all_objects`, `folder_objects`, `label_objects`

**Comment by norkans7** [comment]
Looks good

**Comment by nicpottier** [comment]
Looks like you need to merge master in to resolve conflicts, but let's get this out today.

**Comment by nicpottier** [comment]
Less coverage!??

**Pull request merged**